src/notevennexus/validators.py

```diff
@@ -0,0 +1,269 @@
+    def __init__(self) -> None:
+        super().__init__(
+            "non_numeric_dataset_has_units",
+            "Non-numeric dataset should not have units attribute",
+        )
+
+    def applies_to(self, node: Dataset | Group) -> bool:
+        return isinstance(node, Dataset) and not np.issubdtype(node.dtype, np.number)
+
+    def validate(self, node: Dataset | Group) -> Violation | None:
+        if 'units' in node.attrs:
+            return Violation(node.name)
+
+
+def is_transformation(node: Dataset | Group) -> bool:
+    return 'transformation_type' in node.attrs
+
+
+class transformation_offset_units_missing(Validator):
+    def __init__(self) -> None:
+        super().__init__(
+            "transformation_offset_units_missing",
+            "Transformation with offset attr should also have offset_units attr.",
+        )
+
+    def applies_to(self, node: Dataset | Group) -> bool:
+        return (
+            isinstance(node, Dataset)
+            and is_transformation(node)
+            and 'offset' in node.attrs
+        )
+
+    def validate(self, node: Dataset | Group) -> Violation | None:
+        if 'offset_units' not in node.attrs:
+            return Violation(node.name)
+
+
+class transformation_depends_on_missing(Validator):
+    def __init__(self) -> None:
+        super().__init__(
+            "transformation_depends_on_missing",
+            "Transformation should have depends_on attribute",
+        )
+
+    def applies_to(self, node: Dataset | Group) -> bool:
+        return isinstance(node, Dataset) and is_transformation(node)
+
+    def validate(self, node: Dataset | Group) -> Violation | None:
+        if 'depends_on' not in node.attrs:
+            return Violation(node.name)
+
+
+physical_components = [
+    'NXaperture',
+    'NXattenuator',
+    'NXbeam',
+    'NXbeam_stop',
+    'NXbending_magnet',
+    'NXcapillary',
+    'NXcollimator',
+    'NXcrystal',
+    'NXdetector',
+    'NXdetector_module',
+    'NXdisk_chopper',
+    'NXfermi_chopper',
+    'NXfilter',
+    'NXflipper',
+    'NXfresnel_zone_plate',
+    'NXgrating',
+    'NXguide',
+    'NXinsertion_device',
+    'NXmirror',
+    'NXmoderator',
+    'NXmonitor',
+    'NXmonochromator',
+    'NXpinhole',
+    'NXpolarizer',
+    'NXpositioner',
+    'NXsample',
+    'NXsensor',
+    'NXslit',
+    'NXsource',
+    'NXvelocity_selector',
+    'NXxraylens',
+]
+
+
+class depends_on_missing(Validator):
+    def __init__(self) -> None:
+        super().__init__(
+            'depends_on_missing',
+            'Group describes a physical component but has no depends_on',
+        )
+
+    def applies_to(self, node: Dataset | Group) -> bool:
+        if isinstance(node, Group):
+            if node.attrs.get('NX_class') in physical_components:
+                return True
+        return False
+
+    def validate(self, node: Dataset | Group) -> Violation | None:
+        if 'depends_on' not in node.children:
+            return Violation(node.name)
```

Comment:
I started wondering if each validator should return a different sub-type of `Violation`, but I don't think you would gain much (apart from maybe making testing easier/more complete), and you would then have to add a new violation type every time you want to add a new validator.

Since the sole goal is to compile a summary of results, and not use the violations any further, it's probably fine as is.